Download artifacts from remote repositories on-demand (hover, completion, validation...)`

When trying to validate or get assistance on some plugin or dependency that's not available in local repo and that's necessary to have locally for smart results, it should be downloaded locally so the LS would then provide smartest results.

Today while working on a side-project, one of the dependencies was unable to resolve.
m2e gave an error when running "update project" (which usually resolves pom dependencies). 
I thought it'd be worth pasting the error here as it might give insight to how m2e downloads artifacts from remote repositories:
```

Failed to read artifact descriptor for com.github.goxr3plus:java-stream-player:jar:10.0.0

org.eclipse.aether.resolution.ArtifactDescriptorException: Failed to read artifact descriptor for com.github.goxr3plus:java-stream-player:jar:10.0.0
	at org.apache.maven.repository.internal.DefaultArtifactDescriptorReader.loadPom(DefaultArtifactDescriptorReader.java:255)
	at org.apache.maven.repository.internal.DefaultArtifactDescriptorReader.readArtifactDescriptor(DefaultArtifactDescriptorReader.java:171)
	at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.resolveCachedArtifactDescriptor(DefaultDependencyCollector.java:530)
	at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.getArtifactDescriptorResult(DefaultDependencyCollector.java:515)
	at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency(DefaultDependencyCollector.java:403)
	at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.processDependency(DefaultDependencyCollector.java:356)
	at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.process(DefaultDependencyCollector.java:344)
	at org.eclipse.aether.internal.impl.collect.DefaultDependencyCollector.collectDependencies(DefaultDependencyCollector.java:247)
	at org.eclipse.aether.internal.impl.DefaultRepositorySystem.collectDependencies(DefaultRepositorySystem.java:279)
	at org.apache.maven.project.DefaultProjectDependenciesResolver.resolve(DefaultProjectDependenciesResolver.java:169)
	at org.apache.maven.project.DefaultProjectBuilder.resolveDependencies(DefaultProjectBuilder.java:225)
	at org.apache.maven.project.DefaultProjectBuilder.build(DefaultProjectBuilder.java:198)
	at org.apache.maven.project.DefaultProjectBuilder.build(DefaultProjectBuilder.java:122)
	at org.eclipse.m2e.core.internal.embedder.MavenImpl.readMavenProject(MavenImpl.java:630)
	at org.eclipse.m2e.core.internal.project.registry.DefaultMavenDependencyResolver.resolveProjectDependencies(DefaultMavenDependencyResolver.java:63)
	at org.eclipse.m2e.core.internal.project.registry.ProjectRegistryManager.refreshPhase2(ProjectRegistryManager.java:551)
	at org.eclipse.m2e.core.internal.project.registry.ProjectRegistryManager.lambda$2(ProjectRegistryManager.java:515)
	at org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.executeBare(MavenExecutionContext.java:177)
	at org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.execute(MavenExecutionContext.java:151)
	at org.eclipse.m2e.core.internal.project.registry.ProjectRegistryManager.refresh(ProjectRegistryManager.java:513)
	at org.eclipse.m2e.core.internal.project.registry.ProjectRegistryManager.refresh(ProjectRegistryManager.java:365)
	at org.eclipse.m2e.core.internal.project.registry.ProjectRegistryManager.refresh(ProjectRegistryManager.java:312)
	at org.eclipse.m2e.core.internal.project.ProjectConfigurationManager.updateProjectConfiguration0(ProjectConfigurationManager.java:405)
	at org.eclipse.m2e.core.internal.project.ProjectConfigurationManager$2.call(ProjectConfigurationManager.java:352)
	at org.eclipse.m2e.core.internal.project.ProjectConfigurationManager$2.call(ProjectConfigurationManager.java:1)
	at org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.executeBare(MavenExecutionContext.java:177)
	at org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.execute(MavenExecutionContext.java:151)
	at org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.execute(MavenExecutionContext.java:99)
	at org.eclipse.m2e.core.internal.embedder.MavenImpl.execute(MavenImpl.java:1370)
	at org.eclipse.m2e.core.internal.project.ProjectConfigurationManager.updateProjectConfiguration(ProjectConfigurationManager.java:349)
	at org.eclipse.m2e.core.ui.internal.UpdateMavenProjectJob.runInWorkspace(UpdateMavenProjectJob.java:77)
	at org.eclipse.core.internal.resources.InternalWorkspaceJob.run(InternalWorkspaceJob.java:42)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: org.eclipse.aether.resolution.ArtifactResolutionException: Failure to transfer com.github.goxr3plus:java-stream-player:pom:10.0.0 from https://jitpack.io was cached in the local repository, resolution will not be reattempted until the update interval of jitpack.io has elapsed or updates are forced. Original error: Could not transfer artifact com.github.goxr3plus:java-stream-player:pom:10.0.0 from/to jitpack.io (https://jitpack.io): NullPointerException
	at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:423)
	at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts(DefaultArtifactResolver.java:225)
	at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifact(DefaultArtifactResolver.java:202)
	at org.apache.maven.repository.internal.DefaultArtifactDescriptorReader.loadPom(DefaultArtifactDescriptorReader.java:240)
	... 32 more
Caused by: org.eclipse.aether.transfer.ArtifactTransferException: Failure to transfer com.github.goxr3plus:java-stream-player:pom:10.0.0 from https://jitpack.io was cached in the local repository, resolution will not be reattempted until the update interval of jitpack.io has elapsed or updates are forced. Original error: Could not transfer artifact com.github.goxr3plus:java-stream-player:pom:10.0.0 from/to jitpack.io (https://jitpack.io): NullPointerException
	at org.eclipse.aether.internal.impl.DefaultUpdateCheckManager.newException(DefaultUpdateCheckManager.java:226)
	at org.eclipse.aether.internal.impl.DefaultUpdateCheckManager.checkArtifact(DefaultUpdateCheckManager.java:192)
	at org.eclipse.aether.internal.impl.DefaultArtifactResolver.gatherDownloads(DefaultArtifactResolver.java:564)
	at org.eclipse.aether.internal.impl.DefaultArtifactResolver.performDownloads(DefaultArtifactResolver.java:482)
	at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:400)
	... 35 more
	pom.xml	/sample-manager	line 1	Maven Dependency Problem

```

--- a/lemminx-maven/src/test/java/org/eclipse/lemminx/maven/test/RemoteRepositoryTest.java
+++ b/lemminx-maven/src/test/java/org/eclipse/lemminx/maven/test/RemoteRepositoryTest.java
@@ -9,7 +9,10 @@
 package org.eclipse.lemminx.maven.test;
 
 import static org.eclipse.lemminx.maven.test.MavenLemminxTestsUtils.createDOMDocument;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
 
+import java.io.File;
 import java.io.IOException;
 import java.net.URISyntaxException;
 import java.util.Collections;
@@ -17,6 +20,7 @@ import java.util.List;
 import java.util.concurrent.ExecutionException;
 
 import org.eclipse.lemminx.dom.DOMDocument;
+import org.eclipse.lemminx.maven.MavenPlugin;
 import org.eclipse.lemminx.services.XMLLanguageService;
 import org.eclipse.lemminx.settings.SharedSettings;
 import org.eclipse.lsp4j.CompletionItem;
@@ -86,6 +90,26 @@ public class RemoteRepositoryTest {
  	 		Thread.sleep(500);
  		} while (!(((MarkupContent) hover.getContents().getRight()).getValue().contains(description)));
  		// if got out of the loop without timeout, then test is PASSED
+	}
+	
+	@Test(timeout=15000)
+ 	public void testDownloadArtifactOnHover() throws IOException, InterruptedException, ExecutionException, URISyntaxException {
+		languageService.initializeIfNeeded();
+		File mavenRepo = MavenPlugin.getRepositorySystemSession().getLocalRepository().getBasedir();
+		File artifactDirectory = new File(mavenRepo, "org/glassfish/jersey/project/2.19");
+		String description = "Jersey is the open source (under dual CDDL+GPL license) JAX-RS 2.0 (JSR 339)";
+		final DOMDocument document = createDOMDocument("/pom-remote-artifact-download-hover.xml", languageService);
+		final Position position = new Position(14, 18);
+		assertFalse(artifactDirectory.exists());
+ 		Hover hover;
+ 		do {
+ 	 		hover = languageService.doHover(document, position, new SharedSettings());
+ 	 		Thread.sleep(500);
+ 		} while (hover == null);
+ 		
+ 		assertTrue(artifactDirectory.exists());
+ 		assertTrue(artifactDirectory.listFiles().length > 0);
+ 		assertTrue(hover.getContents().getRight().getValue().contains(description));
 	}
 	
 	@Test(timeout=15000)

--- a/lemminx-maven/src/test/resources/pom-remote-artifact-download-hover.xml
+++ b/lemminx-maven/src/test/resources/pom-remote-artifact-download-hover.xml
@@ -1,0 +1,20 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+	<modelVersion>4.0.0</modelVersion>
+
+	<groupId>org.test</groupId>
+	<artifactId>test</artifactId>
+	<version>0.0.1-SNAPSHOT</version>
+	<packaging>pom</packaging>
+
+
+	<dependencies>
+	<dependency>
+		<groupId>org.glassfish.jersey</groupId>
+		<artifactId>project</artifactId>
+		<version>2.19</version>
+	</dependency>
+	</dependencies>
+
+</project>

